### PR TITLE
vae examples fixes

### PR DESCRIPTION
**Pull request by roatienza**
This PR fixes the ff:
- Modular encoder, decoder and VAE resulting to a more concise and easy to follow implementation
- Option to use MSE (if image Gaussian distribution) or Binary Cross Entropy (if pixel-level Bernoulli distribution) for the reconstruction loss
- Plots for viz latent code vs label and latent code vs generated digit have the same scale for easier understanding of the relation between the 2 plots
- MLP implementation hyperparam is closer to the original paper of Kingma
- CNN conv and deconv implementation parameters reduced from 3.4M to just 134k

### CODE REVIEW

File: examples/variational_autoencoder.py [hunk 1]
```diff
@@ -1,111 +1,205 @@
-'''This script demonstrates how to build a variational autoencoder with Keras.
+'''Example of VAE on MNIST dataset using MLP
 
- #Reference
+This VAE has a modular design. The encoder, decoder and vae
```
Comment (fchollet) [inline examples/variational_autoencoder.py]: VAE

File: examples/variational_autoencoder.py [hunk 2]
```diff
@@ -1,111 +1,205 @@
...
+are 3 models that share weights. After training vae,
```
Comment (fchollet) [inline examples/variational_autoencoder.py]: the VAE

File: examples/variational_autoencoder.py [hunk 3]
```diff
@@ -1,111 +1,205 @@
...
+the encoder can be used to  generate latent vectors.
+The decoder can be used to generate MNIST digits by sampling the
+latent vector from a gaussian dist with mean=0 and std=1.
```
Comment (fchollet) [inline examples/variational_autoencoder.py]: Gaussian distribution

File: examples/variational_autoencoder.py [hunk 4]
```diff
@@ -1,111 +1,205 @@
...
 
- - Auto-Encoding Variational Bayes
-   https://arxiv.org/abs/1312.6114
+[1] Kingma, Diederik P., and Max Welling.
```
Comment (fchollet) [inline examples/variational_autoencoder.py]: Prefer the previous `# References` formatting

File: examples/variational_autoencoder.py [hunk 5]
```diff
@@ -1,111 +1,205 @@
...
-
-# note that "output_shape" isn't necessary with the TensorFlow backend
-z = Lambda(sampling, output_shape=(latent_dim,))([z_mean, z_log_var])
-
-# we instantiate these layers separately so as to reuse them later
-decoder_h = Dense(intermediate_dim, activation='relu')
-decoder_mean = Dense(original_dim, activation='sigmoid')
-h_decoded = decoder_h(z)
-x_decoded_mean = decoder_mean(h_decoded)
+    """Implements reparameterization trick by sampling
```
Comment (fchollet) [inline examples/variational_autoencoder.py]: Please format all docstrings in this CL like other docstrings in the codebase:

```python
"""One-line summary ending in a period.

# Arguments
    argument_1: Description (may include type if relevant).
    argument_2: Description.

# Returns

Description of the output.
"""
```

File: examples/variational_autoencoder.py (hunk 1 above)
Comment (roatienza) [reply examples/variational_autoencoder.py]: fixed.

File: examples/variational_autoencoder.py (hunk 2 above)
Comment (roatienza) [reply examples/variational_autoencoder.py]: fixed.

File: examples/variational_autoencoder.py (hunk 3 above)
Comment (roatienza) [reply examples/variational_autoencoder.py]: fixed

File: examples/variational_autoencoder.py (hunk 4 above)
Comment (roatienza) [reply examples/variational_autoencoder.py]: fixed but I retained the authors

File: examples/variational_autoencoder.py (hunk 5 above)
Comment (roatienza) [reply examples/variational_autoencoder.py]: fixed

File: examples/variational_autoencoder.py [hunk 6]
```diff
@@ -1,111 +1,208 @@
...
+                 model_name="vae_mnist"):
+    """Plots labels and MNIST digits as function of 2-dim latent vector
+
+    Arguments:
+        models (tuple): encoder and decoder models
+        data (tuple): test data and label
+        batch_size (int): prediction batch size
+        model_name (string): which model is using this function
+
+    Returns:
```
[the remark this answers is not in the document]
Comment (roatienza) [reply examples/variational_autoencoder.py]: Fixed. Thanks.

File: examples/variational_autoencoder.py [hunk 7]
```diff
@@ -1,111 +1,208 @@
...
- #Reference
+The VAE has a modular design. The encoder, decoder and VAE
+are 3 models that share weights. After training the VAE model,
+the encoder can be used to  generate latent vectors.
+The decoder can be used to generate MNIST digits by sampling the
+latent vector from a Gaussian distribution with mean=0 and std=1.
 
- - Auto-Encoding Variational Bayes
-   https://arxiv.org/abs/1312.6114
+#Reference
```
[the remark this answers is not in the document]
Comment (roatienza) [reply examples/variational_autoencoder.py]: Fixed. Thanks.

File: examples/variational_autoencoder.py [hunk 8]
```diff
@@ -1,111 +1,208 @@
...
-h = Dense(intermediate_dim, activation='relu')(x)
-z_mean = Dense(latent_dim)(h)
-z_log_var = Dense(latent_dim)(h)
+# reparameterization trick
+# instead of sampling from Q(z|X), sample eps = N(0,I)
+# z = z_mean + sqrt(var)*eps
+def sampling(args):
+    """Reparameterization trick by sampling fr an isotropic unit Gaussian.
 
+    Arguments:
```
[the remark this answers is not in the document]
Comment (roatienza) [reply examples/variational_autoencoder.py]: Fixed. Thanks.

File: examples/variational_autoencoder.py [hunk 9]
```diff
@@ -1,111 +1,208 @@
...
+        args (tensor): mean and log of variance of Q(z|X)
 
-def sampling(args):
-    z_mean, z_log_var = args
-    epsilon = K.random_normal(shape=(K.shape(z_mean)[0], latent_dim), mean=0.,
-                              stddev=epsilon_std)
-    return z_mean + K.exp(z_log_var / 2) * epsilon
+    Returns:
```
[the remark this answers is not in the document]
Comment (roatienza) [reply examples/variational_autoencoder.py]: Fixed. Thanks.

File: examples/variational_autoencoder.py [hunk 10]
```diff
@@ -1,111 +1,208 @@
...
+    return z_mean + K.exp(0.5 * z_log_var) * epsilon
+
+
+def plot_results(models,
+                 data,
+                 batch_size=128,
+                 model_name="vae_mnist"):
+    """Plots labels and MNIST digits as function of 2-dim latent vector
+
+    Arguments:
```
[the remark this answers is not in the document]
Comment (roatienza) [reply examples/variational_autoencoder.py]: Fixed. Thanks.
